aspnetcore/fundamentals/http-context/samples/6.x/HttpContextInBackgroundThread/EmailService.cs

```diff
@@ -1,31 +1,42 @@
 namespace HttpContextInBackgroundThread;
 
+public interface IEmailService
+{
+    void SendEmail(string email);
+}
+
 public class EmailService : IEmailService
 {
     private readonly IHttpContextAccessor _httpContextAccessor;
     private readonly ILogger<EmailService> _logger;
-    private const string? BackgroundCorrelationId = "background-correlation-id";
+    private const string UserAgent = "Unknown";
 
     public EmailService(IHttpContextAccessor httpContextAccessor, ILogger<EmailService> logger)
     {
         _httpContextAccessor = httpContextAccessor;
         _logger = logger;
     }
 
-    public void SendEmail(string email)
+    // The userAgent should come directly from the caller to the service if possible.
+    // An explicit parameter makes the API more usable outside of the request flow, is
+    // better for performance and is easier to reason about than relying on ambient state.
+    public void SendEmail(string email, string? userAgent = null)
     {
-        // Services should account for the possibility of HttpContext being null if not called from a request thread.
-        // e.g. When a service is called from a BackgroundWorker the HttpContext is null.
-        // The service should either throw or gracefully handle this.
+        var request = _httpContextAccessor.HttpContext?.Request;
+        var userAgent = request?.Headers["user-agent"].ToString();
+        if (string.IsNullOrEmpty(userAgent))
+        {
+            userAgent = "Unkown";
+        }
 
-        var correlationId = _httpContextAccessor.HttpContext?.Request.Headers["X-Correlation-Id"].ToString() ?? BackgroundCorrelationId;
-
-        _ = SendEmailCoreAsync(correlationId);
+        await _taskQueue.QueueBackgroundWorkItemAsync(cancellationToken => SendEmailCoreAsync(userAgent, cancellationToken));
```

Comment:
A lot of the code suggestions I made cause the code to no longer compile because they rely on changes elsewhere. For example, we'd need a `_taskQueue` field of type `IBackgroundTaskQueue` like the one in https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-6.0&tabs=visual-studio#queued-background-tasks

Reply:
I'll take care of it and mention you if I get stuck.